src/Language/Pirouette/PlutusIR/SMT.hs

```diff
@@ -83,103 +96,296 @@ trPIRFun :: P.DefaultFun -> [PureSMT.SExpr] -> Maybe PureSMT.SExpr
+    continueWith "Data_List" args
+  branchesBuiltinTerm P.IData _ args =
+    continueWith "Data_I" args
+  branchesBuiltinTerm P.BData _ args =
+    continueWith "Data_B" args
+
+  -- pattern matching and built-in matchers
+
+  -- they take the arguments in a different order
+  branchesBuiltinTerm P.ChooseList _ 
+    (TyArg a : TyArg tyR : TermArg lst : TermArg caseNil : TermArg caseCons : excess) = 
+      continueWithListMatch a tyR lst caseNil caseCons excess
+  branchesBuiltinTerm P.ChooseUnit _ (tyR : unit : rest) = 
+    continueWith "Unit_match" (unit : tyR : rest)
+  branchesBuiltinTerm P.ChooseData _ 
+    (tyR : dat : TermArg caseC : TermArg caseM : TermArg caseL : TermArg caseI : TermArg caseB : excess) =
+      continueWithDataMatch dat tyR caseC caseM caseL caseI caseB excess
+
+  -- built-in matchers
+
+  branchesBuiltinTerm P.FstPair _ [tyA@(TyArg a), tyB@(TyArg b), tuple] =
+    continueWith "Tuple2_match"
+      [ tyA, tyB, tuple, tyA
+      , TermArg $ Lam (Ann "x") a $ Lam (Ann "y") b $ App (Bound (Ann "x") 1) [] ]
+  branchesBuiltinTerm P.SndPair _ [tyA@(TyArg a), tyB@(TyArg b), tuple] =
+    continueWith "Tuple2_match"
+      [ tyA, tyB, tuple, tyB
+      , TermArg $ Lam (Ann "x") a $ Lam (Ann "y") b $ App (Bound (Ann "y") 0) [] ]
+
+  branchesBuiltinTerm P.HeadList _ [TyArg a, TermArg lst] =
+    continueWithListMatch a a lst errorTerm (App (Bound (Ann "x") 1) []) []
+  branchesBuiltinTerm P.TailList _ [TyArg a, TermArg lst] =
+    continueWithListMatch a (listOf a) lst errorTerm (App (Bound (Ann "xs") 0) []) []
+  branchesBuiltinTerm P.NullList _ [TyArg a, TermArg lst] =
+    continueWithListMatch a (builtin PIRTypeBool) lst
+      (K $ PIRConstBool True) (K $ PIRConstBool False) []  
+
+  branchesBuiltinTerm P.UnConstrData _ [d] =
+    continueWithDataMatch 
+      d (TyArg (builtin (PIRTypePair (Just PIRTypeInteger) (Just PIRTypeData))))
+      (App (Free $ TermSig "Tuple2") [TermArg $ App (Bound (Ann "i") 1) [], TermArg $ App (Bound (Ann "ds") 0) []]) 
+      errorTerm errorTerm errorTerm errorTerm []
+  branchesBuiltinTerm P.UnMapData _ [d] =
+    continueWithDataMatch 
+      d (TyArg (listOf (builtin (PIRTypePair (Just PIRTypeData) (Just PIRTypeData)))))
+      errorTerm (App (Bound (Ann "es") 0) []) errorTerm errorTerm errorTerm []
+  branchesBuiltinTerm P.UnListData _ [d] =
+    continueWithDataMatch 
+      d (TyArg (listOf (builtin PIRTypeData)))
+      errorTerm errorTerm (App (Bound (Ann "ds") 0) []) errorTerm errorTerm []
+  branchesBuiltinTerm P.UnIData _ [d] =
+    continueWithDataMatch 
+      d (TyArg (builtin PIRTypeInteger))
+      errorTerm errorTerm errorTerm (App (Bound (Ann "i") 0) []) errorTerm []
+  branchesBuiltinTerm P.UnBData _ [d] =
+    continueWithDataMatch 
+      d (TyArg (builtin PIRTypeByteString))
+      errorTerm errorTerm errorTerm errorTerm (App (Bound (Ann "b") 0) []) []
+
+  branchesBuiltinTerm _rest _translator _args = 
+    pure Nothing
+
+
+continueWith :: 
```

Comment:
Might be helpful to put a comment what `continueWith*` family of functions does.

Reply:
Done